codesign/lib/src/file_codesign_visitor.dart

```diff
@@ -0,0 +1,111 @@
+// Copyright 2019 The Flutter Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style license that can be
+// found in the LICENSE file.
+
+import 'dart:async';
+import 'package:codesign/codesign.dart';
+import 'package:file/file.dart';
+import 'package:process/process.dart';
+
+/// Interface for classes that interact with files nested inside of [RemoteZip]s.
+abstract class FileVisitor {
+  const FileVisitor();
+
+  Future<void> visitEmbeddedZip(EmbeddedZip file, String parent, String entitlementParentPath);
+  Future<void> visitRemoteZip(RemoteZip file, Directory parent);
+  Future<void> visitBinaryFile(BinaryFile file, String entitlementParentPath);
+}
+
+enum NotaryStatus {
+  pending,
+  failed,
+  succeeded,
+}
+
+/// Codesign and notarize all files within a [RemoteArchive].
+class FileCodesignVisitor extends FileVisitor {
+  FileCodesignVisitor({
+    required this.tempDir,
+    required this.commitHash,
+    required this.processManager,
+    required this.codesignCertName,
+    required this.codesignPrimaryBundleId,
+    required this.codesignUserName,
+    required this.appSpecificPassword,
+    required this.codesignAppstoreId,
+    required this.codesignTeamId,
+    required this.stdio,
+    required this.filepaths,
+    this.production = false,
+  });
+
+  /// Temp [Directory] to download/extract files to.
+  ///
+  /// This file will be deleted if [validateAll] completes successfully.
+  final Directory tempDir;
+
+  final String commitHash;
+  final ProcessManager processManager;
+  final String codesignCertName;
+  final String codesignPrimaryBundleId;
+  final String codesignUserName;
+  final String appSpecificPassword;
+  final String codesignAppstoreId;
+  final String codesignTeamId;
+  final Stdio stdio;
+  final bool production;
+  final List<String> filepaths;
+  // Utility utility = Utility();
+  // TODO(xilaizhang): add back utitlity in later splits
+  Set<String> fileWithEntitlements = <String>{};
+  Set<String> fileWithoutEntitlements = <String>{};
+  Set<String> fileConsumed = <String>{};
+
+  late final File entitlementsFile = tempDir.childFile('Entitlements.plist')
+    ..writeAsStringSync(_entitlementsFileContents);
+
+  late final Directory remoteDownloadsDir = tempDir.childDirectory('downloads')..createSync();
+  late final Directory codesignedZipsDir = tempDir.childDirectory('codesigned_zips')..createSync();
+
+  int _remoteDownloadIndex = 0;
+  int get remoteDownloadIndex => _remoteDownloadIndex++;
+
+  static const String _entitlementsFileContents = '''
+<?xml version="1.0" encoding="UTF-8"?>
```

Comment:
Using something like https://pub.dev/packages/mustache_template would be cleaner

Reply:
umm this will be a constant string and unmodified. I might be wrong, but looking at mustache's template it is good for dynamic args?